src/log/file_appender.cpp

```diff
@@ -82,13 +82,13 @@ namespace fc {
                    out.flush();
                    out.close();
                }
+               string rez;
+               read_file_contents(link_filename, rez);
```

Comment:
I don't think this is the correct solution. Log files can be very very large, for example, hundreds of Gigabytes.

Reply:
@abitmore what exactly do you mean ? read_file_contents isn't a good solution ?

Reply:
Correct.

Reply:
ok sure understood, thanks a lot ! :+1:

Reply:
will do it right now :-)

Reply:
@abitmore default behaviour was next: default.log file was erased each rotation time and all its data log inside this file was in default.log.TIMESTAMP file, so a huge 100GB data log always was separated in different files while default.log is hard link for the latest new created dfault.log.TIMESTAMP file. So do we need log file that will be huge 100GB ? Really cannot understand exactly why default behaviour isn't correct and what exactly do we need to achieve as a right result ? @abitmore @jmjatlanta @ryanRfox @oxarbitrage @clockworkgr guys who can help to understand this issue ?

Reply:
remove_all(link_filename);  // on windows, you can't delete the link while the underlying file is opened for writing
               if( fc::exists( log_filename ) )
                  out.open( log_filename, std::ios_base::out | std::ios_base::app );
               else
                  out.open( log_filename, std::ios_base::out | std::ios_base::app);

               create_hard_link(log_filename, link_filename);

This is the code that was before my PR.

Reply:
so I mean that maybe current live version is correct ? Because we don't lose any log data in current live version. No data is overwritten, all logs are save just in separated log files by timestamp string at the end of file name. @ryanRfox @jmjatlanta @oxarbitrage @abitmore @clockworkgr guys what do you think about this issue ?

Reply:
@abitmore @oxarbitrage @ryanRfox @jmjatlanta @clockworkgr guys who can help please ?

Reply:
I believe the original problem was that on a restart, the whatever.log file was getting overwritten. Instead of grabbing its contents and copying it into a new whatever.log file, we want to open the whatever.log file in append mode. Opening in append mode will not overwrite the file. If it gets overwritten, I would guess that something else is causing it (perhaps the hard link has something to do with it?).

Now with a test written, it should be easier to test these scenarios and make sure the contents of the log files contain what we want...
- Starting a log when a whatever.log file exists
- Starting a log when a whatever.log.{timestamp} exists and we should continue with that file
- Starting a log when a whatever.log.{timestamp} exists but should be rolled (note: The hardlink would need to be unlinked in this case)

Reply:
on restart the whatever.log isn't overwritten because as I said earlier each rotation period new whatever.log.timestamp is created and whatever.log file is just hard link which is removed each rotation period and is created again hard link for newest whatever.log.timestamp file, so whatever.log isn't default file, because it's just a hard link, but all logs data is separated in different multiple whatever.log.timestamp.

@jmjatlanta we don't need to grab content from whatever.log file because it is just hard link and the data which is written in whatever.log file will be placed in some whatever.log.timestamp file. So just simple - whatever.log is only hard link for whatever.log.timestamp file.

@jmjatlanta what do you mean we have to do with logs ? do you think we lose some logs data because it's overwritten ? if so, then as I've said earlier we don't lose any logs data, all logs data are in different files whatever.log.timestamp and we won't have any huge files like 100GB in size. So @jmjatlanta @abitmore @oxarbitrage @clockworkgr @ryanRfox what can you say about this issue ? What exactly do we need to implement here ? Still cannot understand why default behaviour that is currently live (before my Pull Request) isn't correct ? Who wrote this code ? Daniel Larimer ? So guys I think IMHO default behaviour was and is correct ! Any thoughts ?

Reply:
cannot understand why we need to open file in append mode so it would be bigger and bigger each new time. Don't you want to have all logs data separated in different files instead of single huge files as now is ? or do you want guys to have 1 single huge file which can be 100GB and greater in it's sizes and at the same time to have the same data logs separated in different whatever.log.{timestamp} files as now is, so do you want mixed solutions ? or what do we need exactly can anybody help me please to understand correctly everything ?

Thanks a lot !

Reply:
Okay, let's make sure we are all thinking the same thing. The original issue said that if you start a witness node with logging set to go to a file appender, the witness node will create a file. If you stop and restart that witness node, what was in the log file originally was overwritten by the new entries.

Are you saying that the original code did not overwrite the log file?

Here is what I think should happen when the witness node starts up:

**Scenario 1** - The witness node is configured to roll the log file every 10 minutes. We start the node at 13:00, and shut it down at 13:02. We should have a log file whatever.log.20180702T130000 as well as a whatever.log. These two files are hard-linked, and contain the same information. If we start the witness node again at 13:04 and stop it again at 13:06, there should still be 2 log files, both with the same information, and both with the information from 13:00 through 13:06.

**Scenario 2** - The witness node is configured to roll the log file every 10 minutes. We start the node at 13:00, and shut it down at 13:02. We should have a log file whatever.log.20180702T130000 as well as a whatever.log. These two files are hard-linked, and contain the same information. If we start the witness node again at 13:30 and stop it again at 13:35, there should now be 3 log files. whatever.log.20180702T130000 that contains log entries from 13:00 through 13:02, whatever.log.20180702T13300000 that contains log entries from 13:30 through 13:35, and whatever.log which is the same as whatever.log.20180702T13300000.

Do we agree that this should happen? Does this happen with the original code?

Reply:
yes the original code works as you have described 2 scenarios. It is correct behaviour IMHO, right ?
So we don't need to change here anything ? right ?

Reply:
I just ran a test with the original code in the develop branch. By default, it is set to roll every hour. I stopped witness_node at 11:22 local time, and verified that I had 2 log files with the same entries until 11:22.

I restarted my witness node, and now have the same 2 files, but the entries start at 11:28 local time. So the original code in the develop branch is broken.

Now I would like to test the changes you made. But before I do, I would like to see the file copy problem fixed. As your new code is written now, a 100GB file will require >100GB RAM to copy. That needs to be fixed.

Reply:
will check again original code, it must work, very strange. So did you see that some logs data was removed completely and we don't see old logs data even in whatever.log.{timestamp} file ?

Reply:
Both files (the timestamp and non-timestamp file) were the same size, with the same information. Old entries were not there.

Reply:
let me check again original code from develop branch

Reply:
so if this is broken then I should just fix it without copying file and without making huge files right ?

Reply:
That would be great, and would solve the issue.

Reply:
:+1: Now I understood all things clearly enough and now I know what exactly we expect as a best quality result for this issue ! Thanks a lot for all your efforts !!! :+1: